homeassistant/components/alexa/smart_home.py

```diff
@@ -1276,6 +1356,158 @@ def async_api_previous(hass, config, request, entity):
     return api_message(request)
 
 
+def api_error_temp_range(request, temp, min_temp, max_temp, unit):
+    """Create temperature value out of range API error response.
+
+    Async friendly.
+    """
+    temp_range = {
+        'minimumValue': {
+            'value': min_temp,
+            'scale': API_TEMP_UNITS[unit],
+        },
+        'maximumValue': {
+            'value': max_temp,
+            'scale': API_TEMP_UNITS[unit],
+        },
+    }
+
+    msg = 'The requested temperature {} is out of range'.format(temp)
+    _LOGGER.info(msg)
+    return api_error(
+        request,
+        error_type='TEMPERATURE_VALUE_OUT_OF_RANGE',
+        error_message=msg,
+        payload={'validRange': temp_range},
+    )
+
+
+def temperature_from_object(temp_obj, to_unit, absolute=True):
+    """Get temperature from Temperature object in requested unit."""
+    from_unit = TEMP_CELSIUS
+    temp = float(temp_obj['value'])
+
+    if temp_obj['scale'] == 'FAHRENHEIT':
+        from_unit = TEMP_FAHRENHEIT
+    elif temp_obj['scale'] == 'KELVIN':
+        # convert to Celsius if absolute temperature
+        if absolute:
+            temp -= 273.15
+
+    if not absolute:
+        if from_unit != to_unit:
+            if from_unit == TEMP_FAHRENHEIT:
+                return temp / 1.8
+            return temp * 1.8
+        return temp
+
+    return convert_temperature(temp, from_unit, to_unit)
+
+
+@HANDLERS.register(('Alexa.ThermostatController', 'SetTargetTemperature'))
+@extract_entity
+@asyncio.coroutine
+def async_api_set_target_temp(hass, config, request, entity):
+    """Process a set target temperature request."""
+    unit = entity.attributes[CONF_UNIT_OF_MEASUREMENT]
+    min_temp = entity.attributes[climate.ATTR_MIN_TEMP]
```

Comment:
Just a note, and this should be checked before we get into this method, but if a state is unavailable, these attributes do not exist.